### PR TITLE
Fixed union/intersection write types

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/56922

cc @andrewbranch

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -14674,7 +14674,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         else {
             result.links.type = isUnion ? getUnionType(propTypes) : getIntersectionType(propTypes);
             if (writeTypes) {
-                result.links.writeType = isUnion ? getUnionType(writeTypes) : getIntersectionType(writeTypes);
+                result.links.writeType = isUnion ? getIntersectionType(writeTypes) : getUnionType(writeTypes);
```
Comment (Andarist) [inline src/compiler/checker.ts:14677]: after [the original change](https://github.com/microsoft/TypeScript/pull/47674) has been merged it [has been mentioned in the comments](https://github.com/microsoft/TypeScript/pull/47674#discussion_r935892404) that likely this should work in the way I have adjusted it here but I don't see any open issue that would be the result of this discussion there

Comment (craigphicks) [reply src/compiler/checker.ts:14677]: 1. Currently the getters and setter seem to be converted right away to plain readonly property (for getter) and plain property (for setter).  So will write types exist for setters at this point?
2. What about the case where `if (propTypes.length > 2){ ... }` before this `else` clause?  Isn't that branch  handled later in different code which also has to be updated?

Comment (Andarist) [reply src/compiler/checker.ts:14677]: 1. if I understand you correctly - what is displayed depends on the location. In write locations, the write type is displayed and in read locations, you get the read type. To change this the TS team would have to receive a well-motivated feature request with many examples of how it would differ from the current behavior.
2. Yes - you are right. I'll take a look at this.
3. What do you consider a plain type here? Could you rephrase this question?

**Comment by craigphicks** [comment]
Although this is labelled as "fixes #56894" it does not change the behavior of the test case listed in that issue.

[See here](https://github.com/microsoft/TypeScript/pull/56915#discussion_r1438938034)

**Comment by craigphicks** [comment]
An unintuive [difference in read behavior between a getter with a setter, and a getter without a setter](https://github.com/microsoft/TypeScript/pull/56915#discussion_r1438938644).

It seems to me like not treating getter and setters as functions means you end up having to creating exceptions in order to get them to behave like properties - and in the end they are neither properties or functions, but a instead a third class.

You would feel it more directly if you had to create the user documentation for it.

**Comment by Andarist** [comment]
I think that part of what you mention here is likely related to [this](https://github.com/microsoft/TypeScript/issues/56894#issuecomment-1871903003) and I'd treat this as another issue. If you prefer, I can remove the reference to your issue from this PR.

**Comment by craigphicks** [comment]
> I think that part of what you mention here is likely related to [this](https://github.com/microsoft/TypeScript/issues/56894#issuecomment-1871903003) and I'd treat this as another issue. If you prefer, I can remove the reference to your issue from this PR.

Of course I prefer if this pull (appropriately modified) fixes #56894!!!  
I posted #56922 which supersedes #56894 to highlight the runtime error risk.

**Comment by Andarist** [comment]
I changed the referenced issue here as I don't want to fix the variance measurement for setters as part of this PR.

**Comment by andrewbranch** [comment]
@typescript-bot perf test this
@typescript-bot user test this
@typescript-bot test top200

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Review by gabritto — APPROVED** [review]

**Comment by gabritto** [comment]
As pointed out already, we should either not close #56922 with this PR or, preferably, re-open #56894 to track the setter variance computation issue.

**Comment by Andarist** [comment]
I think the currently linked issue can be closed with this. From my PoV, its main focus is on those unsound writes that are fixed here. I agree that the other issue could be reopened to track the variance computation issue.

**Comment by andrewbranch** [comment]
I'm surprised I can't find discussion on this, but I think we discussed this when we added the divergent accessor type feature and decided to make the implementation consistent with the (obviously incorrect) way that normal properties work. This PR does not make this an error:

```ts
class One {
    x!: string
}
class Two {
    x!: number
}
declare let u1: One | Two;
u1.x = "";
```

I would like to see this fixed everywhere, but that's of course a much bigger and more disruptive change, and I'm not sure if it makes sense to fix one without the other. Even more confusing, the logic changed here doesn't even kick in unless the property has a getter with a _different_ type. This is still not an error:

```ts
declare class OneAccessor {
    set x(value: string);
}
declare class TwoAccessor {
    set x(value: number);
}
declare let u2: OneAccessor | TwoAccessor;
u2.x = "";
```

And it's _still_ not an error when adding getters of the same type. But once one of the getters&mdash;theoretically irrelevant to this assignment&mdash;diverges from its setter, suddenly the assignment is an error? No, unfortunately, I think this is wrong on purpose.

**Comment by Andarist** [comment]
> I would like to see this fixed everywhere, but that's of course a much bigger and more disruptive change, and I'm not sure if it makes sense to fix one without the other.

To confirm - is this something that you want to get fixed? I could certainly try to do this, I'm just looking for a soft confirmation that fixing this is even on the table. As you have said, it could potentially be a big and disruptive change - and I know that sometimes the team errs on the side of backwards compatibility.

**Comment by andrewbranch** [comment]
I'm not sure when it was last discussed, but my impression was it's not on the table. I think the problem is that it's too slippery of a slope down to the realization that all non-`readonly` properties should be strictly invariant, which would be an unacceptably large break without a flag, and possibly too onerous even with a flag.

```ts
let obj: { x: string } | { x: number };
obj.x = ""; // Error, right?
let obj2: { x: unknown } = obj; // ...Uh oh
obj2.x = ""; // Must be ok, so previous line must be error
```

https://github.com/microsoft/TypeScript/issues/14150
https://github.com/microsoft/TypeScript/issues/18770
